### PR TITLE
Performance/canvas selective render

### PR DESCRIPTION
**Pull request by balazsbajorics**
**Problem:**
Whenever EditorState.projectContents changes, the entire canvas re-renders, including all of the user elements, every single time. For larger projects this kills the performance of interactions. Since we aim for a ~60 FPS mouse interaction performance, we need som way to only re-render the parts of the Canvas that actually change during an interaction.

**Fix:**
Standard React memoization of the ComponentRendererComponents is not applicable for two reasons: 
1. it would alter the user's code in a way that might surprise them, especially (only?) if the user code is not yet compatible with react 18 strict mode and expects a re-render in situations where React.Memo would prevent it.
2. More importantly, ComponentRendererComponents use ContextAtoms to access the ProjectContents to be able to update themselves when the underlying projectContext model for the target component changes. 

**TODO**
- [x] ⚠️ dragging Card component in the example project is broken
- [x] ❓ we only update the manipulated instance of a component, and the rest of the instances only update after the drag is released. Is this desirable? If we want all instances to re-render, I should create a useContextAtomSelector for ContextAtoms instead of my current solution.
- [x] 🐛 There seems to be a problem with focusing on components

**Commit Details:**
- added a new `ElementsToRerender` array to editor state
- `usePubSubAtomReadOnly` takes a callback function which decides if the hook's user component should be force-updated when the atom value changes. 
- I use the shouldUpdate callback in ComponentRendererComponent to prevent updates for components which are not on the `ElementsToRerender` list
- ComponentRendererComponent also calls `updateInvalidatedPaths` now (not sure if we need this?)
- new `SetElementsToRerenderCommand` changes the `ElementsToRerender` list
- using `SetElementsToRerenderCommand` in all strategies
- made a for-tests-only action `SET_ELEMENTS_TO_RERENDER` to fix some dom-walker tests

### CODE REVIEW

File: editor/src/core/shared/element-path.ts [hunk 1]
```diff
@@ -271,6 +271,10 @@ export function parentPath(path: ElementPath): ElementPath {
   }
 }
 
+export function isParentComponentOf(maybeParent: ElementPath, maybeChild: ElementPath): boolean {
```
Comment (balazsbajorics) [inline editor/src/core/shared/element-path.ts:274]: I don't remember why I made this function. should I just use `isParentOf`?

File: editor/src/components/canvas/canvas-strategies/keyboard-absolute-resize-strategy.tsx [hunk 2]
```diff
@@ -52,45 +56,46 @@ export const keyboardAbsoluteResizeStrategy: CanvasStrategy = {
...
-            if (key == null) {
-              return []
-            }
-            const drag = getDragDeltaFromKey(key, interactionState.interactionData.modifiers)
-            const edgePosition = getEdgePositionFromKey(key)
-            if (drag.x !== 0 || drag.y !== 0) {
-              return canvasState.selectedElements.flatMap((selectedElement) => {
-                const element = withUnderlyingTarget(
+        commands: interactionState.interactionData.keysPressed.flatMap<
+          AdjustCssLengthProperty | SetElementsToRerenderCommand
```
Comment (seanparsons) [inline editor/src/components/canvas/canvas-strategies/keyboard-absolute-resize-strategy.tsx:60]: Is this type really necessary?

Comment (balazsbajorics) [reply editor/src/components/canvas/canvas-strategies/keyboard-absolute-resize-strategy.tsx]: not really. I removed it in bb5138a9

File: editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx [hunk 3]
```diff
@@ -90,23 +92,40 @@ export function createComponentRendererComponent(params: {
 
     const mutableContext = params.mutableContextRef.current[params.filePath].mutableContext
 
-    const { topLevelElements, imports } = useGetTopLevelElementsAndImports(params.filePath)
-    const { code, highlightBounds } = useGetCodeAndHighlightBounds(params.filePath)
+    const shouldUpdate = () =>
+      !isFeatureEnabled('Canvas Selective Rerender') ||
+      ElementsToRerenderGLOBAL.current === 'rerender-all-elements' ||
+      ElementsToRerenderGLOBAL.current.findIndex((er) => {
```
Comment (seanparsons) [inline editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx:98]: Wouldn't `ElementsToRerenderGLOBAL.current.some(...` make more sense?

File: editor/src/core/shared/element-path.ts (hunk 1 above)
Comment (seanparsons) [reply editor/src/core/shared/element-path.ts:274]: It _looks_ equivalent to `isParentOf`.

File: editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx (hunk 3 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx]: fixed in 1fd1b4c

File: editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx [hunk 4]
```diff
@@ -90,23 +92,43 @@ export function createComponentRendererComponent(params: {
 
     const mutableContext = params.mutableContextRef.current[params.filePath].mutableContext
 
-    const { topLevelElements, imports } = useGetTopLevelElementsAndImports(params.filePath)
-    const { code, highlightBounds } = useGetCodeAndHighlightBounds(params.filePath)
+    const instancePath: ElementPath | null = tryToGetInstancePath(instancePathAny, pathsString)
+
+    const shouldUpdate = () =>
```
Comment (seanparsons) [inline editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx:97]: Can this be declared as a `function` without costing us anything?

**Review by seanparsons — APPROVED** [review]

**Review by gbalint — APPROVED** [review]

File: editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx (hunk 4 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx]: yep, pushed

**Pull request merged**